diff -ur a/cherrypy/lib/covercp.py b/cherrypy/lib/covercp.py
--- a/cherrypy/lib/covercp.py	2016-06-05 15:06:39.000000000 -0600
+++ b/cherrypy/lib/covercp.py	2017-10-10 14:22:12.033334104 -0600
@@ -293,11 +293,11 @@
             root = os.path.dirname(cherrypy.__file__)
         self.root = root
 
-    @cherrypy.exposed
+    @cherrypy.expose
     def index(self):
         return TEMPLATE_FRAMESET % self.root.lower()
 
-    @cherrypy.exposed
+    @cherrypy.expose
     def menu(self, base="/", pct="50", showpct="",
              exclude=r'python\d\.\d|test|tut\d|tutorial'):
 
@@ -355,7 +355,7 @@
                 buffer = []
                 yield template % (lineno, cgi.escape(line))
 
-    @cherrypy.exposed
+    @cherrypy.expose
     def report(self, name):
         filename, statements, excluded, missing, _ = self.coverage.analysis2(
             name)
diff -ur a/cherrypy/lib/cpstats.py b/cherrypy/lib/cpstats.py
--- a/cherrypy/lib/cpstats.py	2016-06-05 15:06:39.000000000 -0600
+++ b/cherrypy/lib/cpstats.py	2017-10-10 14:22:53.009295746 -0600
@@ -475,7 +475,7 @@
         },
     }
 
-    @cherrypy.exposed
+    @cherrypy.expose
     def index(self):
         # Transform the raw data into pretty output for HTML
         yield """
