diff -up yum-utils-1.1.30/needs-restarting.py.old yum-utils-1.1.30/needs-restarting.py
--- yum-utils-1.1.30/needs-restarting.py.old	2014-05-05 14:19:00.000000000 +0200
+++ yum-utils-1.1.30/needs-restarting.py	2014-05-05 14:19:24.251964474 +0200
@@ -81,19 +81,20 @@ def get_open_files(pid):
     try:
         maps = open(smaps, 'r')
     except (IOError, OSError), e:
-        print "Could not open %s" % smaps
+        print >>sys.stderr, "Could not open %s" % smaps
         return files
 
-    for line in maps.readlines():
-        if line.find('fd:') == -1:
+    for line in maps:
+        slash = line.find('/')
+        if slash == -1 or line.find('00:') != -1: # if we don't have a '/' or if we fine 00: in the file then it's not _REALLY_ a file
             continue
         line = line.replace('\n', '')
-        slash = line.find('/')
         filename = line[slash:]
-        filename = filename.replace('(deleted)', '') #only mildly retarded
+        filename = filename.split(';')[0]
         filename = filename.strip()
         if filename not in files:
             files.append(filename)
+    maps.close()
     return files
 
 def main(args):
