diff -up yum-utils-1.1.30/needs-restarting.py.old yum-utils-1.1.30/needs-restarting.py
--- yum-utils-1.1.30/needs-restarting.py.old	2011-01-13 08:00:20.000000000 +0100
+++ yum-utils-1.1.30/needs-restarting.py	2014-04-04 11:33:19.545859621 +0200
@@ -43,6 +43,9 @@ import yum
 import glob
 import stat
 from optparse import OptionParser
+sys.path.insert(0,'/usr/share/yum-cli')
+import utils
+
 
 def parseargs(args):
     usage = """
@@ -108,9 +111,10 @@ def main(args):
     
     needing_restart = []
 
+    boot_time = utils.get_boot_time()
     for pid in return_running_pids(uid=myuid):
         try:
-            pid_start = os.stat('/proc/' + pid)[stat.ST_CTIME]
+            pid_start = utils.get_process_time(int(pid), boot_time)['start_time']
         except OSError, e:
             continue
         found_match = False
