--- tmp/yum-complete-transaction.py	2012-04-26 13:35:11.000000000 +0200
+++ tmp/yum-complete-transaction.py	2013-09-24 18:43:31.333027645 +0200
@@ -21,6 +21,7 @@
 from yum.misc import setup_locale
 
 from utils import YumUtilBase
+from yum.constants import TS_REMOVE_STATES
 
 import logging
 import os
@@ -203,6 +204,12 @@
             if action == 'erase':
                 (e, m, u) = self.rpmdb.matchPackageNames([pkgspec])
                 for pkg in e:
+                    pkgtuple = (pkg.name, pkg.arch, pkg.epoch, pkg.ver, pkg.rel)
+                    # Skip processing if already marked as the one to be removed.
+                    # This will elimate issue with tx_member removal when package 
+                    # marked as: 'ud' out of of "install" journal's entry
+                    if self.tsInfo.getMembersWithState(pkgtuple, TS_REMOVE_STATES):
+                        continue
                     self.remove(po=pkg)
