diff -up yum-utils-1.1.30/plugins/fs-snapshot/fs-snapshot.py.original yum-utils-1.1.30/plugins/fs-snapshot/fs-snapshot.py
--- yum-utils-1.1.30/plugins/fs-snapshot/fs-snapshot.py.original	2012-08-22 10:16:22.777266939 +0900
+++ yum-utils-1.1.30/plugins/fs-snapshot/fs-snapshot.py	2012-08-22 10:16:35.061143459 +0900
@@ -80,7 +80,7 @@ def inspect_volume_lvm(conduit, volume):
         # convert /dev/mapper name to /dev/vg/lv for use with LVM2 tools
         # - 'dmsetup splitname' will collapse any escaped characters
         p = Popen(["/sbin/dmsetup", "splitname", "--separator", "/",
-                   "--noheadings",
+                   "--noheadings", "-c",
                    "-o", "vg_name,lv_name", device], stdout=PIPE, stderr=PIPE)
         err = p.wait()
         if err:
