commit 6980fd03590ad51826e9e1da3b4b0cd312d0e49a
Author: James Antill <james@and.org>
Date:   Fri Apr 12 16:10:58 2013 -0400

    Output couldn't find a pkg. for 'foo'. BZ 838158.

diff --git a/debuginfo-install.py b/debuginfo-install.py
index 27d19ac..4215700 100755
--- a/debuginfo-install.py
+++ b/debuginfo-install.py
@@ -156,6 +156,10 @@ class DebugInfoInstall(YumUtilBase):
                                 self.di_try_install(deppo)
                             except yum.Errors.InstallError, e:
                                 self.logger.critical('Could not find debuginfo pkg for dependency package %s' % deppo)
+
+        for pkgname in u:
+            self.logger.critical('Could not find a package for: %s' % pkgname)
+
         #  This is kinda hacky, accessing the option from the plugins code
         # but I'm not sure of a better way of doing it
         if not self.no_debuginfo_plugin and self.tsInfo:
