--- /tmp/reposync.py	2014-01-31 10:07:22.797788498 -0500
+++ yum-utils-1.1.30/reposync.py	2014-01-31 10:10:52.280189310 -0500
@@ -198,6 +198,7 @@
         # maybe this shouldn't be entirely fatal
         sys.exit(1)
     
+    exit_code = 0
     for repo in my.repos.listEnabled():
         reposack = ListPackageSack(my.pkgSack.returnPackages(repoid=repo.id))
 
@@ -277,7 +278,6 @@
 
         download_list.sort(sortPkgObj)
         n = 0
-        exit_code = 0
         for pkg in download_list:
             n = n + 1
             repo = my.repos.getRepo(pkg.repoid)
