From de33e923b6af308ca9f5aec5e5c3f625dc903f48 Mon Sep 17 00:00:00 2001
From: Haikel Guemar <hguemar@fedoraproject.org>
Date: Tue, 3 Mar 2015 15:28:22 +0100
Subject: [PATCH] Old nose compat

---
 setup.cfg | 3 ---
 1 file changed, 3 deletions(-)

diff --git a/setup.cfg b/setup.cfg
index b140696..eeb3f0b 100644
--- a/setup.cfg
+++ b/setup.cfg
@@ -1,9 +1,6 @@
 [nosetests]
-logging-clear-handlers = true
 with-coverage = true
 cover-package = urllib3
-cover-min-percentage = 100
-cover-erase = true
 
 [flake8]
 max-line-length = 99
-- 
2.1.0

