diff -up Python-2.6.8/Objects/stringobject.c._Py_HashSecret_Initialized Python-2.6.8/Objects/stringobject.c
--- Python-2.6.8/Objects/stringobject.c._Py_HashSecret_Initialized	2012-04-12 16:21:10.680088000 -0400
+++ Python-2.6.8/Objects/stringobject.c	2012-04-12 16:21:48.397616464 -0400
@@ -1209,7 +1209,9 @@ string_hash(PyStringObject *a)
     register unsigned char *p;
     register long x;
 
+#ifdef Py_DEBUG
     assert(_Py_HashSecret_Initialized);
+#endif
     if (a->ob_shash != -1)
         return a->ob_shash;
     len = Py_SIZE(a);
diff -up Python-2.6.8/Objects/unicodeobject.c._Py_HashSecret_Initialized Python-2.6.8/Objects/unicodeobject.c
--- Python-2.6.8/Objects/unicodeobject.c._Py_HashSecret_Initialized	2012-04-12 16:21:19.357979510 -0400
+++ Python-2.6.8/Objects/unicodeobject.c	2012-04-12 16:21:53.814548743 -0400
@@ -6692,7 +6692,9 @@ unicode_hash(PyUnicodeObject *self)
     register Py_UNICODE *p;
     register long x;
 
+#ifdef Py_DEBUG
     assert(_Py_HashSecret_Initialized);
+#endif
     if (self->hash != -1)
         return self->hash;
     len = PyUnicode_GET_SIZE(self);
