From 8851de5ac8dabd6431328ecbe614e5e78d994ccd Mon Sep 17 00:00:00 2001
From: Ralph Bean <rbean@redhat.com>
Date: Sat, 10 Oct 2015 14:39:19 -0400
Subject: [PATCH] Pin urllib3 to 1.10.4.

---
 setup.py | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/setup.py b/setup.py
index f98f528..77a370c 100755
--- a/setup.py
+++ b/setup.py
@@ -26,7 +26,9 @@ packages = [
     'requests.packages.urllib3.packages.ssl_match_hostname',
 ]
 
-requires = []
+requires = [
+    'urllib3==1.10.4',
+]
 
 version = ''
 with open('requests/__init__.py', 'r') as fd:
-- 
2.4.3

