diff --git a/Lib/ctypes/util.py b/Lib/ctypes/util.py
index b2c514d..d92af0c 100644
--- a/Lib/ctypes/util.py
+++ b/Lib/ctypes/util.py
@@ -139,7 +139,9 @@ elif os.name == "posix":
             finally:
                 rv = f.close()
             if rv == 10:
-                raise OSError, 'objdump command not found'
+                return os.path.basename(f) #  This is good for GLibc, I think,
+                                           # and a dep on binutils is big (for
+                                           # live CDs).
             res = re.search(r'\sSONAME\s+([^\s]+)', dump)
             if not res:
                 return None
