diff -up pciutils-3.1.10/lib/sysfs.c.sysfsfill pciutils-3.1.10/lib/sysfs.c
--- pciutils-3.1.10/lib/sysfs.c.sysfsfill	2012-05-28 15:16:50.000000000 +0200
+++ pciutils-3.1.10/lib/sysfs.c	2014-05-06 16:00:19.872935376 +0200
@@ -211,6 +211,7 @@ sysfs_fill_slots(struct pci_access *a)
       char namebuf[OBJNAMELEN], buf[16];
       FILE *file;
       unsigned int dom, bus, dev;
+      int res = 0;
       struct pci_dev *d;
 
       /* ".", ".." or a special non-device perhaps */
@@ -229,8 +230,16 @@ sysfs_fill_slots(struct pci_access *a)
       if (!file)
 	continue;
 
-      if (!fgets(buf, sizeof(buf), file) || sscanf(buf, "%x:%x:%x", &dom, &bus, &dev) < 3)
-	a->warning("sysfs_fill_slots: Couldn't parse entry address %s", buf);
+      if (!fgets(buf, sizeof(buf), file) || (res = sscanf(buf, "%x:%x:%x", &dom, &bus, &dev)) < 3)
+	{
+	  /*
+	   * In some cases, the slot is not tied to a specific device before
+	   * a card gets inserted. This happens for example on IBM pSeries
+	   * and we need not warn about it.
+	   */
+	  if (res != 2)
+	    a->warning("sysfs_fill_slots: Couldn't parse entry address %s", buf);
+	}
       else
 	{
 	  for (d = a->devices; d; d = d->next)
